pkg/antctl/commanddefinition.go

```diff
@@ -0,0 +1,408 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package antctl
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"reflect"
+	"strings"
+
+	"github.com/fatih/structtag"
+	"github.com/spf13/cobra"
+	"gopkg.in/yaml.v2"
+	"k8s.io/klog"
+
+	"github.com/vmware-tanzu/antrea/pkg/antctl/handlers"
+)
+
+type formatterType string
+
+const (
+	jsonFormatter  formatterType = "json"
+	yamlFormatter  formatterType = "yaml"
+	tableFormatter formatterType = "table"
+)
+
+// commandGroup is used to group commands, it could be specified in commandDefinition.
+// The default commandGroup of a commandDefinition is `flat` which means the command
+// is a direct sub-command of the root command. For any other commandGroup, the
+// antctl framework will generate a same name sub-command of the root command for
+// each of them, any commands specified as one of these group will need to be invoked
+// as:
+//  antctl <commandGroup> <command>
+type commandGroup uint
+
+const (
+	flat commandGroup = iota
+	get
+)
+
+var groupCommands = map[commandGroup]*cobra.Command{
+	get: {
+		Use:   "get",
+		Short: "Get the status or resource of a topic",
+		Long:  "Get the status or resource of a topic",
+	},
+}
+
+const (
+	// tagKey is the tag name of the antctl specific annotation.
+	// For example:
+	// 	type FooResponse struct {
+	// 		Bar BarType `antctl:"key"`
+	// 	}
+	// If the field is annotated with antctl:"key", the framework assumes this field
+	// could be used to retrieve a unique Response, thus the framework will generate
+	// corresponding arg options to the cobra.Command.
+	tagKey = "antctl"
+	// tagOptionKeyArg is the option for antctl annotation. It tells the antctl
+	// the field is a primary key.
+	tagOptionKeyArg = "key"
+)
+
+// argOption describes one argument which can be used in a RequestOption.
+type argOption struct {
+	name      string
+	fieldName string
+	usage     string
+	key       bool
+}
+
+// commandDefinition defines options to create a cobra.Command for a antctl client.
+type commandDefinition struct {
+	// Cobra related
+	Use     string // The lower value of it will be used as the endpoint path, like: <API group>/<lower(Use)>.
+	Short   string
+	Long    string
+	Example string // It will be filled with generated examples if it is not provided.
+	// commandGroup represents the group of the command.
+	CommandGroup commandGroup
+	// Controller should be true if this command works for antrea-controller.
+	Controller bool
+	// Agent should be true if this command works for antrea-agent.
+	Agent bool
+	// Singleton should be true if the handler always returns single object. The
+	// antctl assumes the response data as a slice of the objects by default.
+	Singleton bool
+	// antctl will not add formatter option for this command if it is true.
```

Comment:
Do we have such cases? In general I hope antctl to consume normal Controller/Agent API which could be potentially consumed by other consumers, and so formatting to CLI output should be done by anctl.

Reply:
I have an ongoing CL for `support-bundle` command which only shows a line of bundle path as result. Although the output of the antctl is not formatted, the server side response is still in JSON format.
On a second thought the plaintext is needed, because formatted results are easier to be parsed. I think I should remove this option.

Reply:
Removed.